### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by zerebubuth**
Added some documentation covering the design choices for OSMLR and some of the details of its implementation. It ended up being a mixture of high level stuff and very low level stuff, which I'm not sure I'm happy with.

@kevinkreiser, @dnesbitt61 please could you have a read through, and let me know what you think?

### CODE REVIEW

File: docs/intro.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+This can be somewhat fuzzy, and there may be more than one match. In these cases, it may require some tie-breaking to figure out which is the best match. Descriptors may not overlap, so it's possible to discard any part of the road network which was already matched to a descriptor. Indeed, descriptors which do not have a good candidate match - or have several - might be best postponed until all other good matches have been made to reduce the number of candidates. In general, a more major road such as a highway would make a better match than a residential road, but the circumstances may vary with the characteristics of the individual dataset.
+
+Many of the descriptors will not be fuzzy, and will closely match a single segment of road. In these cases, optimisations are possible so as to not need to run a full route for each match. Running a route can sometimes be expensive in computer time, so avoiding the it whenever possible is recommended.
```
Comment (kevinkreiser) [inline docs/intro.md]: "avoiding the hit" it should say right?

Comment (zerebubuth) [reply docs/intro.md]: Ooops, no. I was aiming for just "avoiding it"... do you think "avoiding the hit" would be better?

Comment (kevinkreiser) [reply docs/intro.md]: oh i dont have a preference, whichever you like more works for me!

Comment (zerebubuth) [reply docs/intro.md]: Thanks! Fixed in 1cdcd18.

**Comment by drewda** [comment]
This is great! A few suggestions:

- Could benefit from a diagram or two. A contrived roadway network with some LRPs and their properties listed?
- Include an example of an OSMLR ID and its component ingredients?
- As part of the Open Traffic POC development, there was quite a lot of discussion about linear referencing options on https://github.com/opentraffic/architecture/issues/1. Call out that this solves that issue (or briefly identify the parts is solves and doesn't solve)?
- Add a "Relationship to Valhalla" section? Would be useful to identify how OSMLR is coupled with Valhalla and how it isn't.
- Inline links to other repos and Valhalla docs, GH issues for future additions, etc.

**Comment by zerebubuth** [comment]
Thanks for the feedback @drewda!

I've added some diagrams, which hopefully makes it easier to understand what the text attempts to explain, and an example OSMLR descriptor. I've linked to that discussion you mentioned (first time I've seen it - some good points there), and linkified the text a little.

I stopped short of adding in GH issues or any kind of future work. I think that's worthwhile doing when this is incorporated into a blog post, but I feel like it's confusing for documentation to reference future work. It might make sense to have "known issues" and "roadmap" documents, but in my experience these quickly fall out of date and cause problems for anyone trying to understand the software. This isn't so much of a problem for a blog post, as it's normal for a blog post to be written once and not revised, but I think there are expectations that documentation is kept current. I think it would be better to leave known issues and future work in the GH issues.

**Comment by drewda** [comment]
@zerebubuth nice additions! I think the graphics are especially helpful, and the final section sets the stage well for people to think about how this can work with not just traffic but also other observations that they may want to attach to segments/ways.

In terms of wrapping up docs, do the readme's in this doc and the tile spec repo need any updates?

**Comment by drewda** [comment]
@kevinkreiser @dnesbitt61 any other suggestions at this point?

Looks ready to me.

**Comment by kevinkreiser** [comment]
@drewda nope id say :rocket:

**Pull request merged**